### PR TITLE
notifier for governance proposals just open for voting

### PR DESCRIPTION
**Pull request by microwavedcola1**
Here is an attempt at the discord notifier, the simple script scans for mgno proposals opened in last 5 mins, and notifies to a WEBHOOK_URL of choice

### CODE REVIEW

File: scripts/governance-notifier.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  const realmProposals = Object.fromEntries(
+    Object.entries(proposals)
+      .filter(
+        ([_k, v]) =>
+          Object.keys(realmGovernances).includes(
+            v.info.governance.toBase58()
+          ) && v.info.votingAtSlot
+      )
+      .sort(
+        (a, b) => b[1].info.votingAt.toNumber() - a[1].info.votingAt.toNumber()
```
Comment (mschneider) [inline scripts/governance-notifier.ts:70]: you might not need this sort

File: scripts/governance-notifier.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    )
+  )
+
+  const realmProposals = Object.fromEntries(
+    Object.entries(proposals)
+      .filter(
+        ([_k, v]) =>
+          Object.keys(realmGovernances).includes(
+            v.info.governance.toBase58()
+          ) && v.info.votingAtSlot
```
Comment (mschneider) [inline scripts/governance-notifier.ts:67]: the last filter condition is also not needed anymore

File: scripts/governance-notifier.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import { PublicKey } from '@solana/web3.js'
+import axios from 'axios'
+import { RealmInfo } from '../@types/types'
+import { getAccountTypes, Governance, Proposal } from '../models/accounts'
+import { getGovernanceAccounts, pubkeyFilter } from '../models/api'
+import { ParsedAccount } from '../models/serialisation'
+import { ENDPOINTS } from '../stores/useWalletStore'
+
+global.fetch = require('node-fetch')
```
Comment (mschneider) [inline scripts/governance-notifier.ts:9]: image

File: scripts/governance-notifier.ts (hunk 2 above)
Comment (microwavedcola1) [reply scripts/governance-notifier.ts:67]: actually if I remove the `&& v.info.votingAtSlot`, then I also get the draft proposals

File: scripts/governance-notifier.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      fiveMinutesSeconds
+    ) {
+      const msg = `${proposal.info.name} proposal just opened for voting`
+      console.log(msg)
+      axios.post(process.env.WEBHOOK_URL, { msg })
+    }
+  }
+}
+
+runNotifier()
```
Comment (mschneider) [inline scripts/governance-notifier.ts:101]: run with setInterval

File: scripts/governance-notifier.ts (hunk 1 above)
Comment (microwavedcola1) [reply scripts/governance-notifier.ts:70]: fixed

File: scripts/governance-notifier.ts (hunk 3 above)
Comment (microwavedcola1) [reply scripts/governance-notifier.ts:9]: I would like to re-use the code from the governance-ui, hence this ugly statement 😅

**Review by mschneider — APPROVED** [review]

File: scripts/governance-notifier.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      // not yet signed i.e. only in draft
+      !proposal.info.signingOffAt
+    ) {
+      countSkipped++
+      continue
+    }
+
+    if (
+      // proposal opened in last 5 mins
+      nowInSeconds - proposal.info.signingOffAt.toNumber() <=
```
Comment (SebastianBor) [inline scripts/governance-notifier.ts:91]: Voting starts when `votingAt` is set and not `signingOffAt`. `signingOffAt` is set when signatories start signing off a proposal. If you have just one signatory it'is going to be the same timestamp as `votingAt` but it doesn't have to be in general case

File: scripts/governance-notifier.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    )
+  )
+
+  console.log(`- scanning all proposals`)
+  let countJustOpened = 0
+  let countSkipped = 0
+  for (const k in realmProposals) {
+    const proposal = realmProposals[k]
+
+    if (
```
Comment (SebastianBor) [inline scripts/governance-notifier.ts:73]: You can filter out all proposals in pre voting state using `proposal.isPreVotingState()`

File: scripts/governance-notifier.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      !proposal.info.signingOffAt
+    ) {
+      countSkipped++
+      continue
+    }
+
+    if (
+      // proposal opened in last 5 mins
+      nowInSeconds - proposal.info.signingOffAt.toNumber() <=
+      fiveMinutesSeconds
```
Comment (SebastianBor) [inline scripts/governance-notifier.ts:92]: I think it would be good to add some tolerance to accommodate for potential time drift and difference in clocks

**Pull request merged**